### ISSUE TITLE
[Bug]: Audio breaks up when using Bluetooth speaker on iOS

### ISSUE THREAD

**Issue by hollance** [juce-framework/JUCE#1456]
### Detailed steps on how to reproduce the bug

I have an AUv3 that's wrapped inside a standalone app running on iPad. When the app is active and I switch to a Bluetooth speaker, the sound becomes all garbled. This also happens if Bluetooth is active before starting the app. Without Bluetooth, it sounds fine. Other audio apps (not written with JUCE) work fine over Bluetooth. The audio processing code in my AUv3 can handle different sample rates and buffer sizes without problems.

I have traced it to the following: When Bluetooth is active, iOS for some reason first attempts to set the sample rate to 8000 Hz (and use a mono bus). With `JUCE_IOS_AUDIO_LOGGING` enabled, it prints the following:

```
Setting target sample rate: 8000
Actual sample rate: 44100
```

It's possible to reproduce this behavior even without Bluetooth. In `juce_Audio_ios.cpp`, in `open()`, change the line `targetSampleRate = sampleRateWanted;` to `targetSampleRate = 8000.0;`.

After some debugging, I was able to fix the issue by doing the following in `juce_Audio_ios.cpp` in `setTargetSampleRateAndBufferSize()`:

```c++
    void setTargetSampleRateAndBufferSize()
    {
        JUCE_IOS_AUDIO_LOG ("Setting target sample rate: " << targetSampleRate);
        sampleRate = trySampleRate (targetSampleRate);
        if (sampleRate != targetSampleRate)
            sampleRate = trySampleRate (sampleRate);
        JUCE_IOS_AUDIO_LOG ("Actual sample rate: " << sampleRate);

        JUCE_IOS_AUDIO_LOG ("Setting target buffer size: " << targetBufferSize);
        bufferSize = tryBufferSize (sampleRate, targetBufferSize);
        if (bufferSize != targetBufferSize)
            bufferSize = tryBufferSize (sampleRate, bufferSize);
        JUCE_IOS_AUDIO_LOG ("Actual buffer size: " << bufferSize);
    }
```

`trySampleRate` sets the `preferredSampleRate` on the `AVAudioSession`, even if the audio hardware cannot handle that sample rate. My hypothesis is that at some point iOS may still attempt to use this preferred rate, while JUCE is using a different sample rate. My fix/workaround changes the `preferredSampleRate` to something that is actually supported. Same thing for the buffer size.

### What is the expected behaviour?

The audio plays over the Bluetooth speaker.

### Operating systems

iOS

### What versions of the operating systems?

Tried on iPadOS 17.6.1 on iPad Pro (10.5-inch). Several beta testers have mentioned this Bluetooth problem on their devices as well. JUCE 8.0.3.

### Architectures

ARM, 64-bit

### Stacktrace

```shell
n/a
```

### Plug-in formats (if applicable)

AUv3

### Plug-in host applications (DAWs) (if applicable)

Standalone app

### Testing on the `develop` branch

I have not tested against the `develop` branch

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by reuk** [comment]
> I have traced it to the following: When Bluetooth is active, iOS for some reason first attempts to set the sample rate to 8000 Hz (and use a mono bus). With JUCE_IOS_AUDIO_LOGGING enabled, it prints the following:

I'd be interested to see the stack trace that leads to `setTargetSampleRateAndBufferSize` being called with a target sample rate of 8 kHz.

I've tested this a bit, both on the develop branch, and on the 8.0.3 release. I'm using a 9th gen iPad running iOS 17.7. So far, I haven't been able reproduce the behaviour you describe when connecting and disconnecting a bluetooth headset while audio is playing via the AudioPlaybackDemo in the DemoRunner.

Another issue in this area was reported after the 8.0.3 release [here](https://forum.juce.com/t/bug-fix-in-iosaudioiodevice-trybuffersize-for-ios-18/62818/27). My potential fix for that issue is to add the following at line 523 in juce_Audio_ios.cpp, in `updateAvailableSampleRates()` before the call to AudioUnitAddPropertyListener:
```
sampleRate = trySampleRate (sampleRate);
bufferSize = getBufferSize (sampleRate);
```

The idea is to set the stored sampleRate and bufferSize back to whatever values they held before querying the available sample rates. This looks similar to the solution you're suggesting, but in a different location.

Please could you try making the change I outlined above and see whether it solves the problem for you?

**Comment by hollance** [comment]
Sure, I can try your solution tomorrow and set a breakpoint to capture the stack trace.

**Comment by hollance** [comment]
When I connect to Bluetooth (cheap AirPods imitation) while the app is running, the JUCE_IOS_AUDIO_LOGGING stuff prints the following:

```
handleRouteChange: New device available
Updating hardware info
Lowest supported sample rate: 44100
Highest supported sample rate: 44100
Available sample rates: 44100
Sample rate after detecting available sample rates: 44100
Available buffer sizes: 1024
Buffer size after detecting available buffer sizes: 1024
Setting target sample rate: 44100
Actual sample rate: 44100
Setting target buffer size: 256
Actual buffer size: 1024
Input channel configuration: {Number of hardware channels: 2, Hardware channel names: "Left" "Right", Are channels available: yes, Active channel indices:, Inactive channel indices: 0 1}
Output channel configuration: {Number of hardware channels: 2, Hardware channel names: "SL-001 Left" "SL-001 Right", Are channels available: yes, Active channel indices: 0 1, Inactive channel indices:}
Creating the audio unit
Internal buffer size: 4096
```

Note that here the sample rate is fine but there is a mismatch between the buffer sizes. This causes the sound to break up.

The stack trace for `setTargetSampleRateAndBufferSize` in this scenario is as follows:

![Screenshot 2024-10-23 at 11 45 36](https://github.com/user-attachments/assets/f0a2e976-0ef4-418d-9f53-7d547af0ed8a)

By the way, if I turn off the Bluetooth headset while the app is still running, the logging shows:

```
handleRouteChange: Old device unavailable
Updating hardware info
Lowest supported sample rate: 22050
Highest supported sample rate: 48000
Trying a sample rate of 23050, got 24000
Trying a sample rate of 25000, got 32000
Trying a sample rate of 33000, got 44100
Trying a sample rate of 45100, got 48000
Available sample rates: 22050 24000 32000 44100 48000
Sample rate after detecting available sample rates: 44100
Available buffer sizes: 59 118 236 472 944 1888 3763
Buffer size after detecting available buffer sizes: 235
Setting target sample rate: 8000
Actual sample rate: 22050
Setting target buffer size: 256
Actual buffer size: 256
Input channel configuration: {Number of hardware channels: 2, Hardware channel names: "Left" "Right", Are channels available: yes, Active channel indices:, Inactive channel indices: 0 1}
Output channel configuration: {Number of hardware channels: 2, Hardware channel names: "Headphones Left" "Headphones Right", Are channels available: yes, Active channel indices: 0 1, Inactive channel indices:}
Creating the audio unit
Internal buffer size: 4096
```

Note how it suddenly attempts to set the target sample rate to 8000 and chooses the closest, which is 22050, even though this device handles 44100 just fine.

The stacktrace for that is the same as above.

If I now connect the Bluetooth headset again, it might ask for 8000 Hz or for 44100 Hz. It all seems a bit arbitrary. I'm sure that's due to these headphones being a bit janky. ;-)

**Comment by hollance** [comment]
If I connect to Bluetooth before starting the app, the logging shows:

```
Creating iOS audio device
Updating hardware info
Lowest supported sample rate: 8000
Highest supported sample rate: 8000
Available sample rates: 8000
Sample rate after detecting available sample rates: 44100
Available buffer sizes: 44 88 176 352 704 1408 2816 2822
Buffer size after detecting available buffer sizes: 176
Input channel configuration: {Number of hardware channels: 1, Hardware channel names: "SL-001", Are channels available: yes, Active channel indices:, Inactive channel indices: 0}
Output channel configuration: {Number of hardware channels: 1, Hardware channel names: "SL-001", Are channels available: yes, Active channel indices:, Inactive channel indices: 0}
Opening audio device: inputChannelsWanted: 0, outputChannelsWanted: 11, targetSampleRate: 8000, targetBufferSize: 256
Input channel configuration: {Number of hardware channels: 2, Hardware channel names: "Left" "Right", Are channels available: yes, Active channel indices:, Inactive channel indices: 0 1}
Output channel configuration: {Number of hardware channels: 2, Hardware channel names: "SL-001 Left" "SL-001 Right", Are channels available: yes, Active channel indices: 0 1, Inactive channel indices:}
Updating hardware info
Lowest supported sample rate: 44100
Highest supported sample rate: 44100
Available sample rates: 44100
Sample rate after detecting available sample rates: 44100
Available buffer sizes: 64 128 256 512 1024
Buffer size after detecting available buffer sizes: 128
Setting target sample rate: 8000
Actual sample rate: 44100
Setting target buffer size: 256
Actual buffer size: 256
Creating the audio unit
Internal buffer size: 4096
```

Here the sample rate is the issue while the buffer size is OK. Note that it's weird that now it suddenly does support a 256 buffer while previously it only supported 1024.

The stack trace is:

![Screenshot 2024-10-23 at 11 41 32](https://github.com/user-attachments/assets/e009589c-12c6-4924-b4ca-532da518844d)

The App/Window/PluginHolder are essentially copies of the JUCE standalone stuff and it goes wrong with the default standalone as well.

**Comment by hollance** [comment]
> My potential fix for that issue is to add the following at line 523 in juce_Audio_ios.cpp, in updateAvailableSampleRates() before the call to AudioUnitAddPropertyListener:

Unfortunately your suggestion does not work in this case.

**Comment by reuk** [comment]
Hi, just wanted to check whether you'd had a chance to test the latest changes on the develop branch.

**Comment by hollance** [comment]
Sorry, I have not had time to test this yet.
